src/Lexer.js

```diff
@@ -102,6 +103,12 @@ module.exports = class Lexer {
 
     this.blockTokens(src, this.tokens, true);
 
+    // Insert known reflinks into em rules to properly skip over them
+    const rep = Object.keys(this.tokens.links).join('|').replace(/\*/g, '\\*');
```

Comment:
Do we need to escape the links incase they have a regexp character (like `|`)?

I feel like a better way to do this would be to change the em rule to over collect then check for reflinks in the tokenizer.

something like:
1. `inline.em` collects `"*em [ref*link]*"`
2. em tokenizer finds that `"ref*link"` is not in `tokens.links`
3. reduce collection to `"*em [ref*"`

I'm not sure which would be faster but it would be less prone to ReDOS and easier to prevent errors based on reflink text.

Reply:
We do something similar with links to allow nesting an unlimited number of matched parentheses.

Reply:
I will see if I can get this to work. It does look like a "cleaner" way to do it.

Reply:
@UziTech Ah, so I have this working on the example you wrote above, but issues arise in cases like this:
`*theres [not*a*link] here*`
which should render to `<em>theres [not<em>a</em>link] here</em>` because we still need to match the first `*` to the final `*`. So essentially we need to run the em regex again over the same section of text but this time not skipping over the `[ ]` when looking for `*`. This requires a whole separate em regex. And *then*, if it happens to be a line with multiple brackets and some are links and some aren't, it again stops working.
For example `*theres [not*a*link] here but [this*is*alink] here*` should become
`<em>theres [not<em>a</em>link] here but <a href="url">this<em>is</em>alink</a> here</em>`

Reply:
I guess there's also the possibility of an internal asterisk requiring us to *expand* the capture instead of reducing, which will also not work with overcollection. For example:

`*Hello [my *name] is* Joe*`
Which should render to `<em>Hello [my <em>name] is</em> Joe</em>`. The initial capture would skip over the brackets and just get us `*Hello [my *name] is*`, but then by opening up the bracket and finding it is not a link, we now have a mismatched asterisk.

Reply:
My example was just to illustrate that the `inline.em` regex doesn't have to be the only thing we use to determine the end of the `em` token. We can choose how much we "over collect" for example the link regex grabs everything that is not white-space for the href and hands it to the tokenizer to determine the actual end of the href.

The `em` regex could just grab everything up until the last `*`, or the end of the string if no `*` exists past the first character, and let other regexes or functions inside the tokenizer decide where the end should actually be. One thing we will want to watch out for though are slow functions. regex search can be extremely fast but regular expressions can't parse everything (unlimited nesting for example).

Reply:
you can look at commonmarks recommended parsing of emphasis for an idea of what the tokenizer might need to do.

https://spec.commonmark.org/0.29/#-process-emphasis-

Reply:
Right, I hear what you're saying. But the point I'm stuck on is that from what I can see, we still need to detect a pattern that considers the entirety of the sequence, intelligently skipping over or entering into bracketed spaces as needed, all at once. Individually looking over bracketed blocks and checking for known links is simple enough, but then we still need to look at the text as a whole, since skipping one asterisk and not another will propagate changes to the pairing of `em` spans all the way to the end.

I will keep thinking on this though.

Reply:
Moved the logic into Tokenizer and just overwrite any found reflinks with the string `[aaaaa]` of equal length to hide any em delimiters inside. Then run the em regex over the new string; all remaining `*` and `_` are valid, so we get a good result. Then just use the positions of this capture to slice the original string to restore the reflinks.

Hope this is more what you had in mind. :)